contracts/tokenomics/vesting/src/contract.rs

```diff
@@ -320,26 +326,126 @@ fn compute_available_amount(current_time: u64, vesting_info: &VestingInfo) -> St
-            let passed_time = current_time.min(end_point.time) - sch.start_point.time;
-            let time_period = end_point.time - sch.start_point.time;
-            if passed_time != 0 && time_period != 0 {
-                let release_amount = Uint128::from(passed_time).multiply_ratio(
-                    end_point.amount.checked_sub(sch.start_point.amount)?,
-                    time_period,
-                );
-                available_amount = available_amount.checked_add(release_amount)?;
-            }
-        }
+        let unlocked_amount = calc_schedule_unlocked_amount(sch, current_time)?;
+        available_amount = available_amount.checked_add(unlocked_amount)?;
     }
 
     available_amount
         .checked_sub(vesting_info.released_amount)
         .map_err(StdError::from)
 }
 
+/// Calculate unlocked amount for particular [`VestingSchedule`].
+/// This function does not consider released amount.
+fn calc_schedule_unlocked_amount(
+    schedule: &VestingSchedule,
+    current_time: u64,
+) -> StdResult<Uint128> {
+    let mut available_amount = schedule.start_point.amount;
+
+    if let Some(end_point) = &schedule.end_point {
+        let passed_time = current_time.min(end_point.time) - schedule.start_point.time;
+        let time_period = end_point.time - schedule.start_point.time;
+        if passed_time != 0 && time_period != 0 {
+            let release_amount = Uint128::from(passed_time).multiply_ratio(
+                end_point.amount.checked_sub(schedule.start_point.amount)?,
+                time_period,
+            );
+            available_amount = available_amount.checked_add(release_amount)?;
+        }
+    }
+
+    Ok(available_amount)
+}
+
+/// Withdraw tokens from active vesting schedule.
+///
+/// Withdraw is possible if there is only one active vesting schedule.
+/// Active schedule's remaining amount must be greater than withdraw amount.
+/// This function terminates current active schedule (updates end_point)
+/// and creates a new one with remaining amount minus withdrawn amount.
+/// Withdrawn amount receiver is either `receiver` or `info.sender`.
+fn withdraw_from_active_schedule(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    account: String,
+    receiver: Option<String>,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let acc = deps.api.addr_validate(&account)?;
+    let mut vesting_info = VESTING_INFO.load(deps.storage, &acc)?;
+    let block_time = env.block.time.seconds();
+
+    let mut active_schedules = vesting_info.schedules.iter_mut().filter(|schedule| {
+        if let Some(end_point) = schedule.end_point {
+            block_time >= schedule.start_point.time && block_time < end_point.time
+        } else {
+            block_time >= schedule.start_point.time
+        }
```

Comment:
We should be able to withdraw even if there are any one-time schedules. Also let's allow withdrawing when `block_time < schedule.start_point.time`